static/src/javascripts/projects/commercial/modules/messenger.ts

```diff
@@ -0,0 +1,382 @@
+	type: string;
+	value: {
+		id?: string | null | undefined;
+		slotId?: string | null | undefined;
+		height: number;
+		width: number;
+	};
+};
+
+/**
+ * Callbacks that can be registered to fire when receiving messages from an iframe
+ */
+type ListenerCallback = (
+	/**
+	 * The data payload sent by an iframe, and has type `unknown` because we can't
+	 * predict what the iframe will send. It is the responsibility of the callback
+	 * to obtain a value of the desired type (or fail gracefully)
+	 */
+	specs: unknown | null | undefined,
+	/**
+	 * Non-persistent callbacks can be chained together. This value is the return
+	 * value of the previously fired callback in the chain. It is the responsibility
+	 * of the current callback to either ignore it or use it / pass along
+	 */
+	ret: unknown,
+	/**
+	 * Reference to the iframe that is the source of the message
+	 */
+	iframe?: HTMLIFrameElement | null | undefined,
+) => unknown;
+
+/**
+ * The set of listeners currently registered
+ *
+ * Each message sent by an iframe has a `type` field which indicates the kind of
+ * message e.g. `resize`, `measure-ad-load`. One or more listeners is registered
+ * for each type.
+ */
+type Listeners = Record<
+	string,
+	ListenerCallback | ListenerCallback[] | undefined | null
+>;
+
+/**
+ * Types of functions to register a listener for a given type of iframe message
+ */
+export type RegisterListener = (
+	type: string,
+	callback: ListenerCallback,
+	options?: {
+		window?: WindowProxy;
+		persist?: boolean;
+	},
+) => void;
+
+/**
+ * Types of functions to unregister a listener for a given type of iframe message
+ *
+ */
+export type UnregisterListener = (
+	type: string,
+	callback?: ListenerCallback,
+	options?: {
+		window?: WindowProxy;
+	},
+) => void;
+
+const LISTENERS: Listeners = {};
+let REGISTERED_LISTENERS = 0;
+
+const error405 = {
+	code: 405,
+	message: 'Service %% not implemented',
+};
+const error500 = {
+	code: 500,
+	message: 'Internal server error\n\n%%',
+};
+
+/**
+ * Determine if an unknown payload has the shape of a programmatic message
+ *
+ * @param payload The unknown message payload
+ */
+const isProgrammaticMessage = (
+	payload: unknown,
+): payload is ProgrammaticMessage => {
+	const payloadToCheck = payload as ProgrammaticMessage;
+	return (
+		payloadToCheck.type === 'set-ad-height' &&
+		('id' in payloadToCheck.value || 'slotId' in payloadToCheck.value) &&
+		'height' in payloadToCheck.value
+	);
+};
+
+/**
+ * Convert a legacy programmatic message to a standard message
+ *
+ * Note that this only applies to specific resize programmatic messages
+ * (these include specific width and height values)
+ */
+const toStandardMessage = (
+	payload: ProgrammaticMessage,
+): StandardMessage<{ width: number; height: number }> => ({
+	id: 'aaaa0000-bb11-cc22-dd33-eeeeee444444',
+	type: 'resize',
+	iframeId: payload.value.id,
+	slotId: payload.value.slotId,
+	value: {
+		height: +payload.value.height,
+		width: +payload.value.width,
```

Comment:
Why the + here? Coerce to the value to `number`?

Reply:
🤷  this was already here, I'll have to take a look again!

Reply:
So it looks like you're right and it's used to coerce to `number`. So either we don't need the `+` or I'll have to take a look at the type of `ProgrammaticMessage` again and revise that!

Reply:
Having a look at this again I don't think this coercion is necessary. We don't have messages originating from Rubicon creatives, which means it should just leave the [Appnexus Piggyback](https://github.com/guardian/frontend/blob/clj/messenger-typescript/commercial/app/templates/piggybackResize.scala.js#L7-L8) which sends integers.

Regardless, I'll migrate the Piggyback away from this legacy format separately.